light-prover/prover/batch_append_circuit.go

```diff
@@ -0,0 +1,364 @@
+// - The new root hash of the Merkle tree.
+// - The updated subtrees after insertion.
+func (circuit *BatchAppendCircuit) append(api frontend.API, leaf frontend.Variable, subtrees []frontend.Variable, indexBits []frontend.Variable) (frontend.Variable, []frontend.Variable) {
+	currentNode := leaf
+
+	for i := 0; i < int(circuit.TreeHeight); i++ {
+		isRight := indexBits[i]
+
+		// Update the subtree at this level if we're inserting on the right
+		subtrees[i] = api.Select(isRight, subtrees[i], currentNode)
+
+		// Select the sibling node
+		sibling := api.Select(isRight, subtrees[i], circuit.getZeroValue(api, i))
+
+		// Arrange nodes for hashing
+		left := api.Select(isRight, sibling, currentNode)
+		right := api.Select(isRight, currentNode, sibling)
+
+		// Hash the nodes to get the parent
+		currentNode = abstractor.Call(api, poseidon.Poseidon2{In1: left, In2: right})
+	}
+
+	return currentNode, subtrees
+}
+
+// incrementBits implements binary addition to increment a number represented as a list of bits.
+// It uses XOR and AND operations to efficiently increment the binary number without using
+// traditional arithmetic operations, which is beneficial in zero-knowledge proof circuits.
+//
+// The function works as follows:
+//  1. It starts with a carry of 1 (equivalent to adding 1 to the number).
+//  2. For each bit, from least to most significant:
+//     a. It XORs the current bit with the carry. This effectively adds the bit and carry
+//     without considering a new carry. (0⊕0=0, 0⊕1=1, 1⊕0=1, 1⊕1=0)
+//     b. It ANDs the original bit with the carry to determine if there should be a carry
+//     for the next bit. (0∧0=0, 0∧1=0, 1∧0=0, 1∧1=1)
+//     c. The result of XOR becomes the new value for the current bit.
+//     d. The result of AND becomes the new carry for the next iteration.
+//  3. This process continues for all bits, resulting in the incremented binary number.
+//
+// Example: Incrementing 0111 (7 in decimal)
+// Initial state: 0111, carry = 1
+// i=0: 1⊕1=0, carry=1∧1=1 -> 0110, carry=1
+// i=1: 1⊕1=0, carry=1∧1=1 -> 0010, carry=1
+// i=2: 1⊕1=0, carry=1∧1=1 -> 1010, carry=1
+// i=3: 0⊕1=1, carry=0∧1=0 -> 1000, carry=0
+// Final result: 1000 (8 in decimal)
+func (circuit *BatchAppendCircuit) incrementBits(api frontend.API, bits []frontend.Variable) []frontend.Variable {
+	carry := frontend.Variable(1)
+	for i := 0; i < len(bits); i++ {
+		// XOR operation implements binary addition without carry
+		newBit := api.Xor(bits[i], carry)
+		// AND operation determines if we need to carry to the next bit
+		carry = api.And(bits[i], carry)
+		bits[i] = newBit
+	}
+	return bits
+}
+
+func (circuit *BatchAppendCircuit) getZeroValue(api frontend.API, level int) frontend.Variable {
+	return frontend.Variable(new(big.Int).SetBytes(ZERO_BYTES[level][:]))
+}
+
+func (circuit *BatchAppendCircuit) createHashChain(api frontend.API, length int, hashes []frontend.Variable) frontend.Variable {
+	if length == 0 {
+		return 0
+	}
```

Comment:
Shouldn't this be if length == 1 return hashes[0] ?

Reply:
We covered this case right on the next line https://github.com/Lightprotocol/light-protocol/pull/1254/files/e391cd08c49f91506fd6eb14053e53e6f8dd7e64#diff-41e7c49b3ad64fa70cbeab12a2333eaa7194f8319e9a5f8ee7b1265c86433208R182